codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -284,194 +345,485 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
+            incrementOutboundFlowControlWindow(initialOutboundStreamWindow);
+            pipeline().fireChannelWritabilityChanged();
         }
 
         @Override
-        public String name() {
-            return frame.name();
+        protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+            if (closed) {
+                throw CLOSED_CHANNEL_EXCEPTION;
+            }
+
+            try {
+                if (sizeEstimatorHandle == null) {
+                    sizeEstimatorHandle = config().getMessageSizeEstimator().newHandle();
+                }
+                for (;;) {
+                    final Object msg = in.current();
+                    if (msg == null) {
+                        break;
+                    }
+                    if (msg == REEVALUATE_WRITABILITY_MESSAGE) {
+                        in.remove();
+                        continue;
+                    }
+
+                    final int bytes = sizeEstimatorHandle.size(msg);
+
+                    /*
+                     * The flow control window needs to be decrement before stealing the message from the buffer (and
+                     * thereby decrementing the number of pending bytes). Else, when calling steal() the number of
+                     * pending bytes could  be less than the writebuffer watermark (=flow control window) and thus
+                     * trigger a  writability change.
+                     *
+                     * This code must never trigger a writability change. Only reading window updates or channel writes
+                      * may change the channel's writability.
+                     */
+                    incrementOutboundFlowControlWindow(-bytes);
+
+                    final ChannelPromise promise = in.steal();
+
+                    if (promise.isCancelled()) {
+                        // Write was cancelled so skip it.
+                        continue;
+                    }
+
+                    // TODO(buchgr): Should we also the change the writability if END_STREAM is set?
+                    try {
+                        if (msg instanceof Http2StreamFrame) {
+                            Http2StreamFrame frame = validateStreamFrame(msg);
+                            frame.stream(stream());
+
+                            if (!firstFrameWritten && !isStreamIdValid(stream().id())) {
+                                if (!(frame instanceof Http2HeadersFrame)) {
+                                    throw new IllegalArgumentException("The first frame must be a headers frame. Was: "
+                                            + frame.name());
+                                }
+                                firstFrameWritten = true;
+                                ctx.write(frame).addListener(new ChannelFutureListener() {
+                                    @Override
+                                    public void operationComplete(ChannelFuture future) throws Exception {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS134)